src/rpc_server.rs

```diff
@@ -3064,6 +3111,64 @@ impl RPC for NeptuneRPCServer {
         Ok(())
     }
 
+    // documented in trait. do not add doc-comment.
+    async fn pow_solution(
+        self,
+        _context: tarpc::context::Context,
+        token: rpc_auth::Token,
+        nonce: Digest,
+    ) -> RpcResult<bool> {
+        log_slow_scope!(fn_name!());
+        token.auth(&self.valid_tokens)?;
+
+        let Some(mut proposal) = self
+            .state
+            .lock_guard()
+            .await
```

Comment:
The lock guard is acquired multiple times. Can this be factored out into a variable?

Reply:
That's intentional. To return as quickly as possible and release all locks as quickly as possible.